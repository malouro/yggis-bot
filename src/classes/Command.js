import noop from 'lodash/noop'

import { MissingPropertyError, InvalidPropertyError } from '../utils/errors'
import { canUseCommand } from '../utils/commands'

export default class Command {
	constructor(options = {}) {
		/* Default options if not provided */
		Object.assign(this, {
			category: 'misc',
			aliases: [],
			disabled: false,
			permLevel: 0,
			description: '',
			usage: { name: this.name || '', args: [] },
		}, options)

		/* Command properties to verify */
		this.validateOptions()
	}

	validateOptions() {
		if (!this.name) {
			throw new MissingPropertyError('name')
		}

		if (
			typeof this.permLevel !== 'number'
			|| !(this.permLevel >= 0 && this.permLevel <= 5)
		) {
			throw new InvalidPropertyError('permLevel', 'Not an integer in range (0, 5)')
		}
	}

	preAction(args) {
		noop(args)
	}

	action(args) {
		noop(args)
	}

	postAction(args) {
		noop(args)
	}

	async run({
		bot,
		client,
		message,
		args,
		logger,
	}) {
		const actionArgs = {
			bot,
			client,
			message,
			args,
			logger,
		}

		if (this.disabled) {
			return null
		}

		if (!canUseCommand(message.member, this.permLevel)) {
			return message.reply(
				'You do not have the required permissions to use this command. ðŸ˜¥'
			)
		}

		await this.preAction(actionArgs)
		await this.action(actionArgs)
		await this.postAction(actionArgs)
	}
}
