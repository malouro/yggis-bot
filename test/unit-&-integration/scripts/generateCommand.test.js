import fs from 'fs'
import path from 'path'
import { exec } from 'child_process'
import { promisify } from 'util'
import waitForExpect from 'wait-for-expect'

import Command from '../../../src/classes/Command'

const asyncExec = promisify(exec)
const getPathToFile = fileName => path.resolve(process.cwd(), fileName)

/**
 * @func generateCommand Runs the script with arguments given
 * @param {Array} args List of CLI arguments to send
 * @param {Object: {stdout, stderror}} callback Callback function that returns the script results
 */
const generateCommand = async (args, callback) => {
	let scriptResults = null

	try {
		scriptResults = await asyncExec(
			'node ./scripts/generate-command.js '
			+ `${args.join(' ')}`,
		)
	} catch (commandError) {
		return callback({ commandError })
	}

	if (callback) {
		return callback(scriptResults)
	}

	return scriptResults
}

/**
 * @func cleanUp Cleanup a generated file after test is done
 * @param {String} fileName Name/path of file to check for and delete
 */
const cleanUp = async (fileName) => {
	const fileToCleanUp = getPathToFile(fileName)

	// eslint-disable-next-line no-bitwise
	fs.access(fileToCleanUp, fs.constants.F_OK | fs.constants.W_OK, (error) => {
		// Error checking
		if (error) {
			if (error.code === 'ENOENT') {
				return 1
			}
			throw error
		}

		// Delete the existing generated .env file & check that it's gone
		fs.unlinkSync(fileToCleanUp)
		expect(() => fs.readFileSync(fileToCleanUp)).toThrowError()
	})

	return 0
}

describe('generate-command', () => {
	const testDir = './test-generations/'
	const pathToTestDir = path.resolve(__dirname, '../../../test-generations')

	beforeAll(async () => {
		fs.stat(pathToTestDir, (err) => {
			if (err) {
				fs.mkdirSync(pathToTestDir)
			}
		})

		await waitForExpect(() => {
			expect(() => fs.statSync(pathToTestDir)).not.toThrow()
		})
	})

	test('makes a command with given options', async () => {
		const testCommandName = 'Test'
		const testOutput = `${testDir + testCommandName}.js`

		await generateCommand([
			`--outputDir="${testDir}"`,
			`--name="${testCommandName}"`,
			'--aliases alias1 alias2 alias3',
		], ({ stdout }) => {
			expect(stdout).toMatchSnapshot('stdout')
		})

		const fileContents = fs.readFileSync(getPathToFile(testOutput), 'UTF8')

		expect(fileContents).toMatchSnapshot('File content')

		await cleanUp(testOutput)
	})

	test('has a command help menu', (done) => {
		generateCommand([
			'--help',
		], ({ stdout }) => {
			expect(stdout).toMatchSnapshot()
			done()
		})
	})

	test('requires a name', (done) => {
		generateCommand([
			`--outputDir="${testDir}"`,
		], ({ commandError }) => {
			expect(commandError.message).toMatch('Command failed')
			expect(commandError.message).toMatch('Missing required argument: n')
			done()
		})
	})

	test('generated file is a usable Yggis Command', async () => {
		const testCommandName = 'TestUsable'
		const testOutput = `${testDir + testCommandName}.js`

		await generateCommand([
			`--outputDir="${testDir}"`,
			`--name="${testCommandName}"`,
			'--aliases alias1 alias2 alias3',
		])

		expect(fs.statSync(getPathToFile(testOutput))).toBeTruthy()

		// eslint-disable-next-line global-require,import/no-dynamic-require
		const CommandToTest = require(getPathToFile(testOutput)).default

		expect(new CommandToTest()).toBeInstanceOf(Command)

		await cleanUp(testOutput)
	})
})
