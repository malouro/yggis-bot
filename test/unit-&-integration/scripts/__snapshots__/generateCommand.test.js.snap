// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `""`;

exports[`generate-command has a command help menu 1`] = `
"Usage: node generate-command.js --

Options:
  --help                     Show help                                 [boolean]
  --version                  Show version number                       [boolean]
  -n, --name, --commandName  Name of the command to make     [string] [required]
  -c, --category             Category/directory for the command
                                               [string] [required] [default: \\"\\"]
  -a, --aliases              List of aliases for the command
                                                           [array] [default: []]
  -o, --outputDir            Filename/path to output the new command to
                                                        [string] [default: \\"./\\"]
"
`;

exports[`generate-command makes a file with expected file content: File content 1`] = `
"import { Command } from '../src/classes'

export default class Test extends Command {
	constructor() {
		super({
			name: 'Test',
			aliases: ['alias1', 'alias2', 'alias3'],
		})
	}

	// eslint-disable-next-line no-unused-vars
	action({ client, message, logger }) {
		// Define the action of the command in here
	}
}
"
`;
